KEYCLOAK-10927 - Implement LDAPv3 Password Modify Extended Operation …

cachescrubber opened the pull request:
…(RFC-3062).

KEYCLOAK-10927 - Implement LDAPv3 Password Modify Extended Operation (RFC-3062).

This PR adds a Configuration Option to the LDAP configuration page to enable or disable (default) the use of the LDAPv3 Password Extended Modify Operation when changing user passwords. Please Review my PR.

Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/906.

I'll update KEYCLOAK-10927 as well.

cachescrubber committed:
KEYCLOAK-10927 - Implement LDAPv3 Password Modify Extended Operation (RFC-3062).

mposolda reviewed (changes requested):
@cachescrubber Thanks for the PR and sorry for the late review.

Instead of introducing another switch to LDAP provider (which already has quite a lot of switches), will it be possible to query LDAP server root DSE and query for the supportedExtensions? In case that passwdModifyOID extension is present, we can automatically assume that the PasswordModifyRequest is used.

More details are in the specification https://tools.ietf.org/html/rfc3062 . Especially see the part, which mentions:
```
Servers SHOULD indicate their support for this extended operation by
   providing PasswdModifyOID as a value of the supportedExtension
   attribute type in their root DSE.
```

It will be nice if you create utility for search available "supportedExtensions" . This can be created in the LDAPOperationManager or LDAPIdentityStore and LDAPConfig can be possibly updated based on the available extensions (so it is needed to query this only once and not at every request or so). If it is not possible to query supported extensions for some reason (EG. unavailable permissions or something like this), then assume that there are not extensions available (which means that the extended PAssword Modify operation won't be used as well).

cachescrubber commented:
Your suggested change should be possible to implement and would be in the sense of RFC-3062. Our LDAP server returns **1.3.6.1.4.1.4203.1.11.1** as supportedExtension using the same credentials as configured in keycloak. I currently have no access to LDAP Server other then OpenLDAP. Could you issue the following search against for example AD, 389 and other?

`ldapsearch  -H ldap://ldap.local -D'cn=keycloak,...' -W -s base '(objectClass=*)' supportedExtension
`

In the context of a passwordChange, it would probably not be too much overhead to query the root dse every time?

Regarding the additional Flag. Don't you think we would need an opt out for servers announcing the extended operation but have a broken implementation or other issues?

```
# extended LDIF
#
# LDAPv3
# base <> (default) with scope baseObject
# filter: (objectClass=*)
# requesting: supportedExtension 
#

#
dn:
supportedExtension: 1.3.6.1.4.1.1466.20037
supportedExtension: 1.3.6.1.4.1.4203.1.11.1
supportedExtension: 1.3.6.1.4.1.4203.1.11.3
supportedExtension: 1.3.6.1.1.8

# search result
search: 2
result: 0 Success

# numResponses: 2
# numEntries: 1

```

mposolda commented:
> Your suggested change should be possible to implement and would be in the sense of RFC-3062. Our LDAP server returns 1.3.6.1.4.1.4203.1.11.1 as supportedExtension using the same credentials as configured in keycloak. I currently have no access to LDAP Server other then OpenLDAP. Could you issue the following search against for example AD, 389 and other?

> ldapsearch -H ldap://ldap.local -D'cn=keycloak,...' -W -s base '(objectClass=*)' supportedExtension

Yes, I have access to MSAD, RHDS and FreeIPA and I can access this. However as you pointed, it can probably easily happen that there is some issue in querying this. I would fallback to the empty list of OID and add some warning to the log in that case. Something similar to this:
```
public List<OID> getLDAPSupportedExtensions() {
    try {
       // Query LDAP here
    } catch (NamingException ne) {
        log.warn("Cant query LDAP Root DSE when obtaining list of supported extensions from LDAP server. Fallback to empty list", ne);
        return Collections.emptyList();
    }
}
```


> In the context of a passwordChange, it would probably not be too much overhead to query the root dse every time?

There can be some other used extensions in the future, which can be called more often. I would call it ideally only once and add it as a field to LDAPConfig or maybe even as a field of the LDAPIdentityStore. I suppose that querying those Root DSE can be done around the time when LDAPIdentityStore is about to be created (Maybe in LDAPIdentityStoreRegistry.createLdapIdentityStore or somewhere here).


> Regarding the additional Flag. Don't you think we would need an opt out for servers announcing the extended operation but have a broken implementation or other issues?

I will try to run your PR against some set of LDAP tests, which is available to the Keycloak team. If it passes, I think we are ok :) If there is some issue/bug reported later, we can try to add the flag, however I hope it will be possible to do it without additional flag and instead detect this automatically. This will be also safer option as the "advanced" password update will be used always when it is available (when flag won't be usually changed as admins don't care if they don't know what the particular flag means).

cachescrubber commented:
I see the point with auto configuration and the need to simplify the keycloak configuration.

But in this particular case, I would rather not fall back to standard password modify when the RootDSE query fails for some reason. Imagine a sysadmin breaks the ACL to query the rootDSE. It will go unnoticed, since not many applications use it anyway. Keycloak would happily change passwords using attribute modify instead of the extended op and we end up with potentially many many clear text credentials in the directory. Not good.

I would suggest the following

1. Keep the Option, but make it an Enum instead of a boolean. on,off,auto. Let **auto** become the default.

- on: use extended password modify
- off: do not use extended password modify
- auto: query the RootDSE and use the extended password modify op when announced as supportedExtension; use the attribute modify otherwise. If, for any reason, the RootDSE query fails, with an error, **refuse to change the password** and return an exception / error to the user.

My suggestion to simplify the configuration would be to present a "Query Server Capabilities" button in the admin-ui. Some settings could then be pre-populated based on the result. What do you think?

I just not sure If I'am the right person to implement the suggested ui changes, because ui is normally not my cup of tea and I already took my 3/4 of the time to create the initial PR to figure out how to add the On-Off switch to the ui ;-)

But I could provide some infrastructure in the context of this PR to do so in the future, for example the mentioned RootDSE query routine.

What do you think?

cachescrubber commented:
Here is a List of known LDAP OIDS and their meaning:

Commit:
KEYCLOAK-10927 - Introduce getLDAPSupportedExtensions(). Use result instead of configuration.

cachescrubber commented:
I committed getLDAPSupportedExtensions() to keep the ball rolling. Let's try to find a good solution to the configuration pattern to use.

mposolda commented:
Thanks. I wanted to avoid more configuration properties as the LDAP configuration page already looks quite complex with so many switches. However what you mention makes sense.

Is it possible to create additional section like "LDAP server extensions" and add the new property to it? It will be nice if this section is hidden by default and available when you click on it - see for example configuration of client details in KEycloak admin console and the hidden sections like "Fine Grain OpenID Connect Configuration" or "Advanced settings". Something similar will be good for the LDAP provider too. If you manage to hide the other existing sections of LDAP provider (Kerberos Integration, Sync Settings, Cache Settings) it will be a nice bonus point and I hope it won't be so much work to do.

Regarding the button, I am not sure if we need it. If the default value will be "auto" as you suggested, then we can have the behaviour like this:
- When LDAP provider is created/updated, the LDAP server will be queried for supported extensions in case that "Password Modify" switch is set to AUTO.
-- If query passes, then everything is fine and the "Advanced Password Modify" can be automatically enabled/disabled based on that
-- If query fails, then creation of LDAP provider will fail and there will be error message described in the admin console or server.log (or both). There will be something like "Failed to query LDAP server for supported LDAP extensions. Please go through 'LDAP server extensions' and switch them manually to ON or OFF"

That way, you won't need to do it so much complicated UI work and deal with the complicated button etc :)

Another bonus point: If you add existing switch "Enable StartTLS" to this "LDAP server extensions" section and use the values like (ON, OFF, Auto) it will be nice. However surely not a blocker for this PR :)

Thanks for your work and suggestions!

cachescrubber commented:
Ok, I'll give it a try.

Regarding StartTLS: This is a good example of a feature which should not activate automatically - wether the server supports it or not. The reason is that with StartTLS the connection pool can't be used and "List User" performance will degrade to a point where Keycloak is no longer usable. See my mailing list issues.

mposolda commented:
@cachescrubber Good point regarding StartTLS. True that it will be better to keep the default as "Off" instead of "Auto" . If you don't change anything regarding StartTLS in this PR, it is completely fine, but whatever you prefer.

Commit:
KEYCLOAK-10927 add possibility to query ldap supported extensions

Commit:
KEYCLOAK-10927: reorganize layout of ldap config page

Commit:
KEYCLOAK-10927 use configuration instead of rootDSE query.

Commit:
KEYCLOAK-10927 Extend the RootDSE query to include controls and features as well as extension OIDs.

cachescrubber commented:
@mposolda We have committed the requested changes to the ui.

regarding the main topic of this PR: The rootDSE query has been moved into the frontend. A new section,  "Advanced Settings" is introduced. Once open, a group of settings appear, next to a Button "Query Supported Extensions".  On button press, a rootDSE query is triggered via REST. The "Enable the LDAPv3 Password Modify Extended Operation" is enabled when the server returns the right OID.  This way, the password feature is independent from the rootDSE stuff.

I have renamed `getLDAPSupportedExtensions` to `queryServerCapabilities`. The call will also request the attributes _supportedControl_, _supportedExtension_ and _supportedFeatures_ and extract their OIDs. I have looked at the KC LDAPConfig and have identified two more settings (StartTLS and Pages Results) which could be discovered by the rootDSE query. I added a comment including the OID in `LDAPConstants`.

The UI changes introduces some scrolling issues. My colleague Kevin has fixed that as well. Since it is a quite massive change to the ui (nearly every page is touched) I think it might be better to track this outside  this PR? What do you think. You can preview the changes in in my forked repo. We can either merge it here, or open another PR.

cachescrubber commented:
@mposolda To be clear about the

mposolda commented on services/src/main/java/org/keycloak/services/resources/admin/RealmAdminResource.java:
```diff
@@ -966,6 +981,43 @@ public Response testLDAPConnection(TestLdapConnectionRepresentation config) {
+        map.add(LDAPConstants.BIND_CREDENTIAL, bindCredential);
+        IdentityStore ldapIdentityStore = new LDAPIdentityStore(session, new LDAPConfig(map));
+
+        Set<LDAPOid> ldapOids = ldapIdentityStore.queryServerCapabilities();
```
It will be better if response allows to differentiate between supportedControl, supportedExtensions and supportedFeatures. The JSON can possibly look like this:
```
{ 
    "supportedExtension": [ "1.3.6.1.4.1.1466.20036" ],
    "supportedFeatures": [ "1.3.6.1.1.14", "1.3.6.1.4.1.4203.1.5.1" ],
    "supportedControl": [ "1.2.840.113556.1.4.1413", "1.2.840.113556.1.4.2239", "1.2.840.113556.1.4.319" ]
}
````
Feel free to create some internal JSON representation class for help with this - for example see TestLdapConnectionRepresentation for the inspiration

mposolda replied:
It will be also good to dump those supported extensions, features and controls to the server.log with the "log.info"

cachescrubber replied:
I though about that too, but my conclusion was it is practically not relevant whether a given OID is a _control_, _extension_ or _feature_. That is way I refactored the method to "capabilities" in the first place. Do you think it is really necessary? The OID is unique anyway. Good point with the info log. Will be added.

mposolda replied:
I think that it may be useful in general to differentiate between control, extension and feature. 

LDAP server itself also differentiates between those 3 things, so from this perspective, differentiation makes sense :)

cachescrubber replied:
Done.

mposolda commented on themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js:
```diff
@@ -1703,7 +1704,23 @@ module.controller('LDAPUserStorageCtrl', function($scope, $location, Notificatio
+            (response) => {
+                Notifications.success("LDAP supported extensions successfully requested.");
+                const ldapOids = response.data;
+                if (angular.isArray(ldapOids)) {
```
This doesn't seem to work for me. My LDAP server has only single supportedExtension "1.3.6.1.4.1.1466.20036" and despite that, when I click on "Query" button, the "Enable the LDAPv3 Password Modify Extended Operation" is incorrectly switched to ON.

Hint: You can run LDAPEmbeddedServer from your IDE, which runs embedded ApacheDS server. That one doesn't have the passwdModifyOID enabled.

cachescrubber replied:
Should be fixed.

mposolda reviewed (changes requested):
@cachescrubber Thanks, nice work!

I've added some inline comments with some minor changes. Besides that, it will be good to have some automated tests. I suggest this:
- It will be good to add new test method to UserFederationLdapConnectionTest to test querying capabilities. This test uses embedded ApacheDS ldap server. You may also need to add new test method to the class org.keycloak.admin.client.resource.RealmResource corresponding to the new REST endpoint you added

Regarding the admin console changes, I don't think that it is necessary. It is surely not something to be included in this PR, but it will eventually need to be separate PR. Feel free to write a mail to keycloak-dev about the proposal around this. Just a note that we are going to rewrite admin console and we try to focus on the new admin console rather than polishing the current admin console and do some major refactoring there.

cachescrubber commented on services/src/main/java/org/keycloak/services/resources/admin/RealmAdminResource.java:
```diff
@@ -966,6 +981,43 @@ public Response testLDAPConnection(TestLdapConnectionRepresentation config) {
+        MultivaluedHashMap<String, String> map = new MultivaluedHashMap<>();
+        map.add(LDAPConstants.CONNECTION_URL, config.getConnectionUrl());
+        map.add(LDAPConstants.USE_TRUSTSTORE_SPI, config.getUseTruststoreSpi());
+        map.add(LDAPConstants.AUTH_TYPE, LDAPConstants.AUTH_TYPE_SIMPLE);
```
That is because` AUTH_TYPE` is not part of `org.keycloak.representations.idm.TestLdapConnectionRepresentation` as you said. I think that the rootDSL query should be done using the same credentials then used later on, to avoid permission problems. So it might be unnecessary to pass AUTH_TYPE, unless something like SASL is added?

mposolda replied:
Yes, I think that it may be needed to either:
1) add authType as another option to TestLdapConnectionRepresentation.
2) assume "bindType" none in case that bindDn and bindCredential are not set?

It seems that (2) might be easier? However the bindType none should be supported. In many cases, administrator sets the bindType "none" when configuring their LDAP provider. In this case, querying the LDAP server rootDSL should be also possible with the bindType "none" .

cachescrubber replied:
Done. I added the authType to TestLdapConnectionRepresentation.

Commit:
KEYCLOAK-10927: fix setting of passwordModifyExtended option

Commit:
KEYCLOAK-10927: refactor LDAPConnectionTestManager and add corresponding test

Commit:
KEYCLOAK-10927: rename LDAPConnectionTestManager to LDAPServerCapabilitiesManager

Commit:
KEYCLOAK-10927: introduce CapabilityType

Commit:
KEYCLOAK-10927: rename Rest Endpoint to ldapServerCapabilities

Commit:
KEYCLOAK-10927: info log discovered capabilities

cachescrubber commented on services/src/main/java/org/keycloak/services/resources/admin/RealmAdminResource.java:
```diff
@@ -966,6 +981,43 @@ public Response testLDAPConnection(TestLdapConnectionRepresentation config) {
+        String bindCredential = handleSavedBindCredential(config.getComponentId(), config.getBindCredential());
+
+        // Create Factory Methods for LDAPConfig and LDAPIdentityStore in LDAPConnectionTestManager?
+        MultivaluedHashMap<String, String> map = new MultivaluedHashMap<>();
```
Done.

cachescrubber commented:
@mposolda I have just committed the following

- add authType to TestLdapConnectionRepresentation
- add org.keycloak.testsuite.admin.UserFederationLdapConnectionTest#testLdapCapabilities
- refactor common code between connection and capabilities test in LDAPServerCapabilitiesManager and RealmAdminResource.
- rename LDAPOid to LDAPCapability, introduce CapabilityType to classify OID.
- Fix user.js

Hopefully this will end the rootDSE excurse and we can get back to business regarding the password change...

mposolda commented:
@cachescrubber Thanks! I am doing some testing in my environment and it seems that when the "Password Modify" switch is enabled and I have "Sync Registrations" ON for my LDAP provider, then I am not able to register new user to the LDAP. It seems the PasswordModify works just when LDAP user already has existing password in the LDAP, however it doesn't work when the LDAP user doesn't already have the existing password. Can we somehow address it? Perhaps use the advanced PasswordModify just in cases when user already has existing password?

cachescrubber commented:
@mposolda Which LDAP Server have you used for your test? I just created a new user using my local OpenLDAP Test server and have been able to change the password without an issue.

What exactly did tried? I did the following steps:

- Users/Add User
- Reset Password

Perhaps in your setup it is necessary to use the _random-userPassword_ mapper?

mposolda committed:
KEYCLOAK-10927 Added LDAPPasswordModifyExtensionTest. A bit of refactoring related to query supported LDAP extensions and PAssword Modify operation

cachescrubber commented:
More test results:

389 DS without TLS refuses to perform the password modify extended operation:

`[LDAP: error code 13 - Operation requires a secure connection.]`

mposolda commented:
@cachescrubber Yes, it is true that random-userPassword mapper helps with this. I've did some changes and test fixes. Please check https://github.com/cachescrubber/keycloak/pull/1 . Thanks!

Commit:
Merge branch 'mposolda-KEYCLOAK-10927_ldap_password_modify_extended_op-2' into KEYCLOAK-10927_ldap_password_modify_extended_op

cachescrubber commented:
@mposolda: Just curious: which LDAP server requires the password attribute to exists before performing a password change? Is it the embedded ApacheDs, or other vendors too?

mposolda reviewed (approved):
@cachescrubber Thanks! ApacheDS for sure, but it is possible that some others as well.

mposolda commented:
@cachescrubber Thanks again for the patience and all your work on this!

cachescrubber commented:
@mposolda Cool. Thanks for your help and support while getting this done!
